keps/sig-scheduling/4187-a-generic-scheduler-cache-extension/README.md

```diff
@@ -0,0 +1,777 @@
+        return err
+    }
+
+    pc. availableCapacity += mapRequest(req)
+    return nil
+}
+```
+
+- Implement the ExtendedResource.Handle which is used to keep the resource's allocable resource update-to-date. Custom Resource(CR) is one of the methods to update the resource allocable resource by a third-party agent or a system admin. And the Handle watches the resource changes and updates the extended resource accordingly. The example uses a self-defined CR, and the CR includes two fields: nodeName and fakeResource. The fakeResource stands for the actual allocable fake resource and nodeName stands for the name of the specific node. 
+``` go
+type Handle struct {
+    cache cache.ExtendedCache
+}
+
+func (fh *Handle) Run(c cache.ExtendedCache, f informers.SharedInformerFactory) error {
+
+    fh.cache = c
+    client, err := versioned.NewForConfig(config)
+    fakeResourceInformerFactory := externalin.NewSharedInformerFactory(client, 0)
+
+    nodeInformer := fakeResourceInformerFactory.Ioi().V1().NodeFakeResourceStatuses().Informer()
+
+    nodeInformer.AddEventHandler(clientcache.ResourceEventHandler{
+        Handler: clientcache.ResourceEventHandlerFuncs{
+            AddFunc:    fh.addResource,
+            UpdateFunc: fh.updateResource,
+            DeleteFunc: fh.deleteResource,
+        },
+    })
+}
+
+func (fh *Handle) updateResource(oldObj, newObj interface{}) {
+    …
+    oldVal := old.Spec.FakeResource
+    newVal := new.Spec.FakeResource
+
+    if oldVal != newVal {
+        res := fh.cache.GetExtendedResource(new.Spec.NodeName)
+        res.(*FakeExtendedResource).SetAvail(newVal)
+        fh.cache.SetExtendedResource(new.Spec.NodeName, res)
+    }
+}
+```
+
+#### Example Implementation of a Scheduler Plugin
+
+A scheduler plugin is to be implemented to initialize the extended resource handles, register them in the registry and fetch the extended resources through the handles at these hook points.
+``` go
+// New initializes a new plugin and returns it.
+func New(_ runtime.Object, h framework.Handle) (framework.Plugin, error) {
+    // initialize the extended resource handles and register them in the registry
+   extendedResourceHandle, err := h.ExtendedResourceHandleRegistry().Register(ResourceType, func() (extendedresource.Handle, error) {
+        return &FakeResourceHandle{}, nil
+    })
+
+   plugin := &FakePlugin{
+        erh: []extendedresource.Handle{extendedResourceHandle},
+    }
+    return plugin, nil
+}
+```
+
+At the filter hook point, the extended resource is retrieved through the extended resource handle that was created during the initialization phase.
+
+```go
+type FakePlugin struct {
+    erh []extendedresource.Handle
+}
+
+func (fp *FakePlugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+    req := ParsePodRequest(pod)
+    …
+
+    nodeName := nodeInfo.Node().Name
+    resource := fp.erh[0].GetExtendedResource(nodeName)
```

Comment:
this is hardcoded as erh[0], is that really to be a slice for `erh`?

Reply:
There is likelihood that a scheduler plugin might make scheduling decision based on multiple resources. In that case, a slice or map would be necessary to manage their handlers.